src/Build/Logging/DistributedLoggers/ConfigurableForwardingLogger.cs

```diff
@@ -151,6 +151,15 @@ private void ApplyParameter(string parameterName)
             {
                 _showCommandLine = true;
             }
+            else if (string.Equals(parameterName, ForwardProjectContentsDescription, StringComparison.OrdinalIgnoreCase)
+                && !_forwardingSetFromParameters)
```

Comment:
Checking `_forwardingSetFromParameters` here means that the logic is now order-dependent. If we hit a `_forwardingTable` key first, `FORWARDPROJECTCONTEXTEVENTS` will be ignored. If we hit `FORWARDPROJECTCONTEXTEVENTS` first, it will be applied. Is this intentional?

Reply:
Mostly! At first I didn't do this, and some tests failed because they were specifying individual events and got more than they bargained for. I refactored to this and justify it like this: if you go to the trouble of specifying individual events we should respect that.

But I'm happy to hear other ideas/prioritizations!

Reply:
Gotcha - and the only way to specify `FORWARDPROJECTCONTEXTEVENTS` is the new code in `CreateForwardingLoggerRecord`? Or is it possible that the logger will see, e.g.

`FORWARDPROJECTCONTEXTEVENTS;LOWMESSAGEEVENT`
and not
`LOWMESSAGEEVENT;FORWARDPROJECTCONTEXTEVENTS`
?

The two would have different behavior which I, as a drive-by reviewer, find suspicious.

Reply:
horribly: no, that's not the only way. You can in fact pass those on the command line, and you can horribly break logging today by doing, for instance:

```sh-session
❯ msbuild -flp:v=diag .\MSBuild.Dev.slnf /clp:PROJECTSTARTEDEVENT /v:d
Microsoft (R) Build Engine version 17.1.0-preview-22055-02+797fd829a for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 1/20/2022 3:50:42 PM.

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:38.98
```

(note no actual output despite fairly high verbosity because most of the events were just dropped)

These are documented in https://docs.microsoft.com/visualstudio/msbuild/writing-multi-processor-aware-loggers. Do you think it would be reasonable to just not document `FORWARDPROJECTCONTEXTEVENTS`? I was hoping that that would get us to the point where we could say with some confidence "only we will add this and we will only ever append it so we can assume it's last". But that's not exactly an iron-clad guarantee.

Reply:
> Do you think it would be reasonable to just not document FORWARDPROJECTCONTEXTEVENTS?

I think it's reasonable. Another option would be to document it, together with its overriding behavior, and tweak the code to just set a flag here and move the actual setting of `_forwardingTable` to `ParseParameters`. There is already some overriding implemented there so it wouldn't look out of place.

https://github.com/dotnet/msbuild/blob/90275e45541681fa48c1acc0fffe4bd23821ce43/src/Build/Logging/DistributedLoggers/ConfigurableForwardingLogger.cs#L116-L122

Reply:
I like that and switched to it.